### PR TITLE
Read data as trustless gateway with car stream

### PR DESCRIPTION
**Pull request by dapplion**
## Summary of changes

Changes introduced in this pull request:
- Download files from IPFS gateway using a car stream to guarantee the integrity of the data. Uses `rs-car-ipfs` library which constructs the file's block DAG in memory and writes to and out AsyncWrite when all nodes are available. This strategy should be sufficient as long as the downloaded files are relatively small.

## Reference issue to close (if applicable)

Closes https://github.com/ChainSafe/forest/issues/2458

## Notes to mantainers
**Gateway safety**

Gateways usually return the content of the file as-is. To ask the gateway to return a CAR stream one must add the query `?format=car`. Since the query is optional old gateways will just return the content as-is which will fail to decode as car.

The trustless gateway feature was shipped a while ago so most providers will support it by now. To test is a gateway supports car streams you can pull the example readme and expect to see binary output
```
curl https://ipfs.io/ipfs/QmQPeNsJPyVWPFDVHb77w8G42Fvo15z4bG2X8D2GhfbSXc?format=car 
```

To test that a specific file can be verified with the underlying car library you can install it as a binary and do
```
cargo install rs-car-ipfs --features bin
curl https://ipfs.io/ipfs/QmQPeNsJPyVWPFDVHb77w8G42Fvo15z4bG2X8D2GhfbSXc?format=car | car-ipfs
```

**Cargo.lock diff**

The libraries `rs-car` and `rs-car-ipfs` depend on some building blocks which the forest dependencies already use like `cid` or `multihash`. Since the dependent versions are different that's the cause of the diff. **Question**: Should I depend on a more flexible version of those deps in my library to allow cargo to resolve it to the same? On proper semver notation `0.8.*` and `0.10.*` and major jumps. Otherwise would you recommend me to downgrade the libraries to using the same set of versions than forest?

## Change checklist

- [x] I have performed a self-review of my own code,
- [x] I have made corresponding changes to the documentation,
- [ ] I have added tests that prove my fix is effective or that my feature works (if possible),
- [ ] I have made sure the [CHANGELOG](https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md) is up-to-date. All user-facing changes should be reflected in this document.

### CODE REVIEW

File: utils/paramfetch/Cargo.toml
Comment (hanabi1224) [inline utils/paramfetch/Cargo.toml:14]: Better to re-export `cid` from `rs-car` in case there is a version mismatch

File: utils/paramfetch/src/lib.rs
Comment (hanabi1224) [inline utils/paramfetch/src/lib.rs:34]: One the files is ~280MB,  10MB might not be sufficient here

**Comment by dapplion** [comment]
Previous implementation piped the HTTP response body to a file, since all bytes are received sequentially thus memory footprint is negligible.

In a CAR stream that may not be the case, since IPLD blocks are de-duplicated. If a latter node in the DAG tree references a previous block of data repeated appearances will not be streamed. If one wants to keep the memory footprint of a CAR streamer low it must have the capability to do random access on any previously streamed data. The naive implementation is to buffer in memory on a hash-map. A better design is possible but requires more engineering complexity. It's doable but would like to know it's really necessary since this issue already took decent involvement.

**Comment by hanabi1224** [comment]
@dapplion Memory usage is not a big deal in the forest-fetch-param case. 

As a side note, if you'd like to extend your lib to support very large files with low memory usage, maybe you could use [parity-db](https://github.com/paritytech/parity-db) which is essentially a hash map on disk

File: utils/paramfetch/Cargo.toml
Comment (dapplion) [reply utils/paramfetch/Cargo.toml]: Done

File: utils/paramfetch/src/lib.rs
Comment (dapplion) [reply utils/paramfetch/src/lib.rs]: Constant removed

**Comment by dapplion** [comment]
@hanabi1224 Switched to `read_single_file_seek` which does not buffer data blocks in memory. If there's a need to recover a previous block it fetches it from the target file. See this comment for reference https://github.com/dapplion/rs-car-ipfs/blob/60a6423f27445315e95c43f8286dbd7f74d4ed6c/src/single_file/single_file_seek.rs#L1-L36

**Comment by hanabi1224** [comment]
@dapplion thanks! LGTM. Will approve once u resolve the merge conflict and have CI passing

File: Cargo.lock
Comment (hanabi1224) [inline Cargo.lock:1315]: @dapplion After thinking about it, I think it would be better if you could relax `cid` version in `rs-car` by using `cid = ">=0.8, <0.11"` in `Cargo.toml`, that way it will automatically resolve to the same version forest uses

Comment (dapplion) [reply Cargo.lock:1315]: Looked into it and `libipld-core` depends on cid 0.10.0. That library contains a cbor decoder which may return Cid data structures from the car stream header. libipld-core v0.14.0.

The API of libipld-core v0.14.0 and v0.16.0 (latest that requires cid v0.10.0) is incompatible. The same source can work for both versions.

So I don't see an easy way to support the relaxed range you suggest

Comment (hanabi1224) [reply Cargo.lock]: That makes sense. Thanks for investigating

**Comment by dapplion** [comment]
@hanabi1224 downgraded the set of deps of rs-car to match this repo. Once you bump to more recent cid and multihash dependencies I can re-bump too

**Comment by hanabi1224** [comment]
@dapplion looks like `rs-car` introduces many unnecessary indirect dependencies, I made a PR to clean them up, please take a look  https://github.com/dapplion/rs-car/pull/8

File: Cargo.lock
Comment (hanabi1224) [inline Cargo.lock:6598]: This crate looks stale

File: Cargo.lock
Comment (hanabi1224) [inline Cargo.lock:7676]: This should not be needed

File: Cargo.lock
Comment (hanabi1224) [inline Cargo.lock:7696]: This should not be needed

File: Cargo.lock
Comment (hanabi1224) [inline Cargo.lock:7913]: This should not be needed

**Comment by dapplion** [comment]
> @dapplion looks like `rs-car` introduces many unnecessary indirect dependencies, I did some investigation and made a PR to clean them up, please take a look [dapplion/rs-car#8](https://github.com/dapplion/rs-car/pull/8)

Thank you for taking a look! Github Actions is currently offline, will merge and release once service is back up

**Comment by dapplion** [comment]
@hanabi1224 upstreamed your changes, lock file looks good now. Did a pass with @ec2 today on both libraries he :+1:

**Review by hanabi1224 — APPROVED** [review]
Nice work, thanks!

**Comment by dapplion** [comment]
Hey @LesnyRumcajs could you provide more details on the error that this PR caused so I can look into it? Including:
- How to reproduce?
- Can you link to the code that does the integrity check?
- Can you link to the set of files / or IPFS hashes so I can download locally?

**Comment by elmattic** [comment]
Hey @dapplion 

I had those errors when trying to sync from scratch. So

1. Make sure you delete the whole Forest directory before.
2. Import a new snapshot calibnet.
3. Run the node.

**Comment by LesnyRumcajs** [comment]
You only need to remove the `filecoin-proof-parameters` directory, no?

**Comment by elmattic** [comment]
I think so, yes. But in my case I was just describing how I have encountered those errors.

**Pull request merged**
